### PR TITLE
Prometheus: Support 1ms resolution intervals

### PR DESCRIPTION
**Pull request by dankeder**
**What this PR does / why we need it**:

Support 1ms timestamps in Prometheus data source and Grafana UI.

**Which issue(s) this PR fixes**:

Possibly related issues: 
- https://github.com/grafana/grafana/issues/3035
- https://github.com/grafana/grafana/issues/7533

**Special notes for your reviewer**:

I discovered that Prometheus datasource is returning wrong timestamps with milliseconds stripped off. This PR should fix that. There is also a change to Grafana UI to make it actually show milliseconds in graphs (where appropriate).

Marking the PR as draft because there are no tests for this yet.

Suggestions/comments are welcome.

### CODE REVIEW

File: packages/grafana-ui/src/components/Graph/utils.ts [hunk 1]
```diff
@@ -118,6 +118,9 @@ export const graphTimeFormat = (ticks: number | null, min: number | null, max: n
     const oneDay = 86400010;
     const oneYear = 31536000000;
 
+    if (secPerTick <= 10) {
+      return systemDateFormats.interval.second.replace('ss', 'ss.SSS');
```
Comment (aocenas) [inline packages/grafana-ui/src/components/Graph/utils.ts:122]: Mayne we could just add `systemDateFormats.interval.milliseconds`

**Comment by danieljosephchambers** [comment]
This PR would solve problems we've had since updating Grafana to 8.3.5 from an older version. We rely heavily on being able to plot Prometheus data at sub-second intervals

File: packages/grafana-ui/src/components/Graph/utils.ts (hunk 1 above)
Comment (dankeder) [reply packages/grafana-ui/src/components/Graph/utils.ts]: Done.

**Comment by dankeder** [comment]
@aocenas Thank you for the review, I changed a couple of things to match your suggestions. I also added a couple of small test cases and changes to the UI so milliseconds are shown if it's appropriate (e.g. in graph tooltips).

Note that most places in the UI will still show only up to seconds resolution. I think this is sufficient as most people probably don't want to see milliseconds by default. Those few of us that do can use the env variable `GF_DATE_FORMATS_FULL_DATE=YYYY-MM-DD HH:mm:ss.SSS` (or the equivalent in grafana.ini) to show milliseconds (in timepicker, graph axis and elsewhere)

**Comment by dankeder** [comment]
Tests are passing now, removing "draft" label.

**Review by axelavargas — COMMENTED** [review]
Hey @dankeder,  great contribution 👏🏾

I reviewed the` format.ts` and `rangeutil.ts` files, code-wise looks good! I think would be great to add the new milliseconds support to the existing unit test `formats.test.ts` and `rangeutils.test.ts`.

File: pkg/tsdb/intervalv2/intervalv2_test.go [hunk 2]
```diff
@@ -69,6 +69,8 @@ func TestRoundInterval(t *testing.T) {
 		interval time.Duration
 		expected time.Duration
 	}{
+		{"10ms", time.Millisecond * 10, time.Millisecond * 1},
+		{"30ms", time.Millisecond * 30, time.Millisecond * 20},
```
Comment (aocenas) [inline pkg/tsdb/intervalv2/intervalv2_test.go:73]: Seems like this is just accidentally duplicated?

**Comment by aocenas** [comment]
Finally tested this and everything seems to work fine. I wonder if we should also make the range selector take less than 1s time diff as this does not work now
Screenshot from 2022-05-02 14-17-23

But that can be left for different PR.

**Comment by aocenas** [comment]
@dankeder how does this look at your end do you have time to look at @axelavargas suggestion? If not we can just take this and finish later.

**Comment by dankeder** [comment]
I'll get to it this week, I hope :pray:

**Comment by sakjur** [comment]
> Finally tested this and everything seems to work fine. I wonder if we should also make the range selector take less than 1s time diff as this does not work now Screenshot from 2022-05-02 14-17-23
> 
> But that can be left for different PR.

We have both a backend and frontend implementation of the range selector, we should make sure that both gets updated when doing this 🙂 I can probably help out with the backend if someone decides to pick this up for the frontend. On the backend side we're using https://github.com/vectordotdev/go-datemath through the `legacydata` (🙃) package and it wouldn't be too hard to add support for millisecond precision in the upstream package as well as long as we remember it.

**Comment by dankeder** [comment]
@axelavargas I updated the tests, please check if it's ok.

As for adding ms support to the parser in time picker, I think it'd be better to leave it for a new PR, it looks like it's needed to change the `go-datemath` package as @sakjur mentions.

File: pkg/tsdb/intervalv2/intervalv2_test.go (hunk 2 above)
Comment (dankeder) [reply pkg/tsdb/intervalv2/intervalv2_test.go]: Sorry, I missed this. It's supposed to be the case for 15ms interval, rounding down to 10ms.

**Review by axelavargas — APPROVED** [review]
Hey @dankeder the tests look great! 🥳 (sorry for the delay), I am approving those changes.

If @aocenas, and @sakjur can approve the rest would be awesome :)

**Review by sakjur — APPROVED** [review]
The Go code looks great :+1:

Would it be a similar PR to add support for nanoseconds or microseconds or would we need more magic to do so? 🤔

**Comment by dankeder** [comment]
Is there anything else that needs to be changed/fixed? Or is it OK and it can be merged as it is?

**Comment by torkelo** [comment]
@dankeder merging this, thanks for your contribution!

File: packages/grafana-data/src/datetime/formats.ts [hunk 3]
```diff
@@ -48,6 +47,15 @@ export class SystemDateFormatsState {
       second: '2-digit',
     });
 
+    // ES5 doesn't support `DateTimeFormatOptions.fractionalSecondDigits` so we have to use
+    // a hack with string replacement.
+    this.fullDateMS = this.fullDate.replace('ss', 'ss.SSS');
```
Comment (aptenodytes-forsteri) [inline packages/grafana-data/src/datetime/formats.ts:52]: Is it possible this is happening even when users *don't* want millisecond precision. For example, I have a table and I have set 

` GF_DATE_FORMATS_FULL_DATE=MMM Do, YYYY HH:mm:ss`.

I see dates of the format `Aug 31st, 2022 17:23:20.067` even though I want `Aug 31st, 2022 17:23:20`

Comment (aptenodytes-forsteri) [reply packages/grafana-data/src/datetime/formats.ts:52]: Nevermind, it appears my version of grafana does not even have this change.

However, similar logic exists on the older tag which I'm still suspicious of: https://github.com/grafana/grafana/blame/v9.0.7/packages/grafana-data/src/datetime/formats.ts#L36

Comment (aptenodytes-forsteri) [reply packages/grafana-data/src/datetime/formats.ts:52]: Commented on https://github.com/grafana/grafana/issues/48917.

I think caused in https://github.com/grafana/grafana/pull/37569

**Pull request merged**